src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/ReleaseProcessorService.cs

```diff
@@ -129,102 +134,120 @@ private static void RemoveSubjectIfExisting(ImportMessage message, StatisticsDbC
-            }
-            statisticsDbContext.SaveChanges();
-            return topic;
+            return CreateOrUpdateWithExclusiveLock(
+                message.Release.Publication.Topic.Title,
+                statisticsDbContext,
+                message.Release.Publication.Topic,
+                message.Release.Publication.Topic.Id,
+                () => new Topic
+                    {
+                        Id = message.Release.Publication.Topic.Id,
+                        Title = message.Release.Publication.Topic.Title,
+                        Slug = message.Release.Publication.Topic.Slug,
+                        Theme = theme
+                    });
         }
 
-        private Theme CreateOrUpdateTheme(ImportMessage message, StatisticsDbContext statisticsDbContext)
+        private Task<Theme> CreateOrUpdateTheme(ImportMessage message, StatisticsDbContext statisticsDbContext)
         {
-            Theme theme;
-            if (!statisticsDbContext.Theme
-                .Any(t => t.Id.Equals(message.Release.Publication.Topic.Theme.Id)))
-            {
-                theme = new Theme
+            return CreateOrUpdateWithExclusiveLock(
+                message.Release.Publication.Topic.Theme.Title,
+                statisticsDbContext,
+                message.Release.Publication.Topic.Theme,
+                message.Release.Publication.Topic.Theme.Id,
+                () => new Theme 
                 {
                     Id = message.Release.Publication.Topic.Theme.Id,
                     Slug = message.Release.Publication.Topic.Theme.Slug,
                     Title = message.Release.Publication.Topic.Theme.Title
-                };
-                theme = statisticsDbContext.Theme.Add(theme).Entity;
-            }
-            else
+                });
+        }
+        
+        /**
+         * This method allows for a concurrency-safe pattern to check for the existence of a given entity and, if it
+         * does not exist to create it, or if it does exist, to update it.
+         *
+         * This is concurrency-safe as it ensures that any two concurrent importer processes that are trying to create
+         * or update the same entity type with the same ID will not be able to do so at the same time.  Rather, the
+         * first thread to reach this code will acquire a lock and prevent the second thread from being able to check
+         * for the existence of the entity until the first thread has finished creating it.  At this point, the second
+         * thread is freed up to check if the entity already exists, which by this point, it will.
+         */
+        private Task<TEntity> CreateOrUpdateWithExclusiveLock<TEntity, TMessageObject>(
+            string entityName,
+            StatisticsDbContext statisticsDbContext,
+            TMessageObject objectFromMessage,
+            Guid entityId,
+            Func<TEntity> createEntityFn)
```

Comment:
Do we need supplier functions for this? Think it'd probably be cleaner (and don't need an additional function context) by just passing in the entities in question

Reply:
Yep, sounds like a good idea.  Previously they had been retrieving related entities from the DB which was why their creation was lazy, but now there's no need for this.